app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/EventsViewModel.kt

```diff
@@ -1,32 +1,142 @@
+    data class Refresh(val isLoggedIn: Boolean) : Action()
+}
+
+sealed class Result {
+    data class EventsLoaded(val events: List<Event>) : Result()
+    object ShowError : Result()
+    object HideError : Result()
+    object None : Result()
+}
+
+class EventsViewModel @Inject constructor(
+        private val eventsLocalRepository: EventsLocalRepository,
+        private val eventsRemoteRepository: EventsRemoteRepository,
+        private val ticketsLocalRepository: TicketsLocalRepository,
+        private val ticketsRemoteRepository: TicketsRemoteRepository,
+        eventType: EventType
 ) : ViewModel() {
 
-    val events: LiveData<List<Event>>
-        get() = when (type) {
-            EventType.ALL -> controller.events
-            else -> controller.bookedEvents
+    private val compositeDisposable = CompositeDisposable()
+    private val refreshRelay = PublishRelay.create<Action>()
+
+    private val _viewState = MutableLiveData<EventsViewState>()
+    val viewState: LiveData<EventsViewState> = _viewState
+
+    init {
+        val initialViewState = EventsViewState.initial()
+
+        val eventsChanges = when (eventType) {
+            EventType.ALL -> eventsLocalRepository.getEvents()
+            EventType.BOOKED -> eventsLocalRepository.getBookedEvents()
         }
 
-    class Factory(
-            private val controller: EventsController,
-            private val eventType: EventType
-    ) : ViewModelProvider.Factory {
+        val databaseChanges = eventsChanges.map(Result::EventsLoaded)
+        val actions = refreshRelay.flatMap(this::processAction)
+
+        compositeDisposable += Observable.merge(databaseChanges, actions)
+                .scan(initialViewState, this::reduceState)
+                .distinctUntilChanged()
+                .subscribe(this::render)
+    }
+
+    private fun processAction(action: Action): Observable<Result> {
+        return when (action) {
+            is Action.Refresh -> fetchEventsAndTickets(action.isLoggedIn)
+        }
+    }
 
-        @Suppress("UNCHECKED_CAST") // no good way around this
-        override fun <T : ViewModel?> create(modelClass: Class<T>): T {
-            return EventsViewModel(controller, eventType) as T
+    private fun reduceState(viewState: EventsViewState, result: Result): EventsViewState {
+        return when (result) {
+            is Result.EventsLoaded -> viewState.toEventsLoaded(result.events)
+            Result.ShowError -> viewState.toError()
+            Result.HideError -> viewState.toNoError()
+            Result.None -> viewState // no changes, we return the current view state
         }
+    }
+
+    private fun fetchEventsAndTickets(isLoggedIn: Boolean): Observable<Result> {
+        return loadAndStoreEvents()
+                .andThen(Observable.just(Result.None as Result))
+                .doOnNext { loadAndStoreTickets(isLoggedIn) }
+                .onErrorResumeNext { t: Throwable -> showErrorForDuration(ERROR_DURATION) }
```

Comment:
I think we can remove the parameter here

Reply:
I could change it to `_`, but removing the parameter results in an ambiguous overload.